File: 013_LeetCode_1/7_complement.py


# Function to find the binary complement of a number

def findComplement(num):
    # Step 1: Convert the number into binary string (remove '0b' prefix)
    binary = bin(num)[2:]  # Example: bin(5) → '0b101' → '101'
    
    # Step 2: Flip the bits manually (change 0 to 1, and 1 to 0)
    flipped = ''  # This will store the flipped binary digits
    for digit in binary:
        if digit == '0':
            flipped += '1'
        else:
            flipped += '0' 
    
    # Step 3: Convert flipped binary string back to decimal (base 2)
    complement = int(flipped, 2)
    
    return complement

n = int(input("Enter a number: "))
print("Complement:", findComplement(n))

File: 013_LeetCode_1/7_complement.cpp
#include<iostream>
using namespace std;

int main() {

    int num, rem, ans=0, mul=1;

    cout << "Enter the Number = ";
    cin >> num;

    if(num==0)
    {
         cout<<"Complement of number is 1 .";
         return 0;
    }

    while (num) {
        rem = num % 2;
        rem=rem^1;
        num = num / 2;
        ans = ans + rem*mul;
        mul*=2;
    }
    cout<<"Complement of number is "<<ans<<" .";

    return 0;

}





/* class Solution {
    public:
    int bitwiseComplement(int n) {

    if(n==0)
         return 1;

         int rem, ans=0, mul=1;
    while (n) 
    {
        rem = n % 2;
        rem=rem^1;
        n = n / 2;
        ans = ans + rem*mul;
        mul*=2;
    }

    return 0;
    }
}; */